// @progress/kendo-schematics/src/collection/grid/files/inline/__name@dasherize__/__name@dasherize__.component.ts
import { Component } from '@angular/core';<% if (dataPath) {%>
import { <%= dataName %> } from '<%= dataPath %>'; <% } %><% if (dataSource === "Example") {%>
import { products } from './data'; <% } %><% if (dataSource === "Mock Data") {%>
import data from './data';<% } %><% if (editing === 'inline' || editing === 'incell') {%>
import { FormBuilder, FormGroup } from '@angular/forms';
import { CreateFormGroupArgs } from '@progress/kendo-angular-grid';<% } %>

@Component({
	selector: '<%= dasherize(name) %>-component',
	template: `
        <kendo-grid <% if (paging || sorting || filtering || grouping) {%>
            [kendoGridBinding]="gridData"<% } else {%>[data]="gridData"<% } %><% if (editing === 'inline') {%>
            [kendoGridReactiveEditing]="createFormGroup"<% }%><% if (editing === 'incell') {%>
            [kendoGridInCellEditing]="createFormGroup"<% }%><% if (filtering){%>
            [filterable]="true"<% }%><% if (grouping) {%>
            [groupable]="true"<% } %><% if (paging) {%>
            [pageable]="true"<% } %><% if (sorting) {%>
            [sortable]="true"<% } %><% if (resizing) {%>
            [resizable]="true"<% } %><% if (reordering) {%>
            [reorderable]="true"<% } %><% if (columnMenu) {%>
            [columnMenu]="true"<% } %><% if (navigable) {%>
            [navigable]="true"<% } %><% if (height) {%>
            [height]="<%=height%>"<% } %><% if (paging) {%>
            [pageSize]="pageSize"><% } %>><% if (columns.length) {%><% for (let column of columns) {%>
            <kendo-grid-column<% if (column.field) {%>
                field="<%=column.field%>"<% } %><% if (column.title) {%>
                title="<%=column.title%>"<% } %><% if (column.hidden) {%>
                [hidden]="<%=column.hidden%>"<% } %><% if (column.locked) {%>
                [locked]="<%=column.locked%>"<% } %><% if (column.sticky) {%>
                [sticky]="<%=column.sticky%>"<% } %><% if (column.width) {%>
                [width]="<%=column.width%>"<% } %><% if (column.columnMenu) {%>
                [columnMenu]="<%=column.columnMenu%>"<% } %><% if (column.editable) {%>
                [editable]="<%=column.editable%>"><% } %>></kendo-grid-column><% } %><% } %><% if (editing === 'inline') {%>
            <kendo-grid-command-column title="Command" [width]="220">
                <ng-template kendoGridCellTemplate let-formGroup="formGroup" let-isNew="isNew">
                    <button kendoGridEditCommand [primary]="true">Edit</button>
                    <button kendoGridRemoveCommand>Remove</button>
                    <button kendoGridSaveCommand [disabled]="formGroup?.invalid">
                        {{ isNew ? "Add" : "Update" }}
                    </button>
                    <button kendoGridCancelCommand>
                        {{ isNew ? "Discard changes" : "Cancel" }}
                    </button>
                </ng-template>
            </kendo-grid-command-column><% } %><% if (editing === 'incell') {%>
            <kendo-grid-command-column title="command" [width]="220">
                <ng-template kendoGridCellTemplate let-formGroup="formGroup" let-isNew="isNew">
                    <button kendoGridRemoveCommand>Remove</button>
                    <button kendoGridSaveCommand [disabled]="formGroup?.invalid">
                    Add
                    </button>
                    <button kendoGridCancelCommand>Discard</button>
                </ng-template>
            </kendo-grid-command-column><% } %><% if (editing === 'incell' || editing === 'inline') {%>
            <ng-template kendoGridToolbarTemplate>
                <button kendoGridAddCommand>Add new</button>
            </ng-template><% } %>
        </kendo-grid>
	`,<% if (inlineStyle) {%>
	styles: [``]<% } else {%>
	styleUrls: ['./<%= dasherize(name) %>.component.css']<% } %>
})
export class <%= classify(name) %>Component  {
	public gridData: any[] = <% if (dataSource === "Example") { %>products<% } else if (dataSource === "Mock Data") { %>data<% } else if (dataName) { %><%=dataName%><% } else { %>products<%} %>;<% if (paging) { %>
	public pageSize: number = <%=pageSize%>;<% if (editing === 'inline' || editing === 'incell') {%>

    constructor(private formBuilder: FormBuilder) { }

    public createFormGroup = (args: CreateFormGroupArgs): FormGroup => {
        // TODO: Implement default item factory.
        const item = args.isNew ? Object.keys(this.gridData[0]).reduce((acc, curr) => (acc as any)[curr] = null || acc, {}) : args.dataItem;

        return this.formBuilder.group(item);
    };<% } %>
}
