// @progress/kendo-schematics/src/collection/grid/schema.ts

export enum DataSource {
  /**
   * An example data source will be added to the project with predefined column definitions.
   * The data and column definitions can not be modified.
   * 
   * This is the default.
   */
  "Example" = "Example",

  /**
   * The user is expected to provide column definitions.
   * A mock data will be generated out of the column definitions.
   */
  "Mock Data" = "Mock Data",

  /**
   * The user can specify a basic module to bind to and provide column definitions,
   * but in general the user will have to author some code to connect the data.
   */
  "Existing App Data" = "Existing App Data"
}

export interface Schema {
    path?: string;
    name: string;
    inlineStyle: boolean;
    inlineTemplate: boolean;
    skipTests: boolean;
    filtering?: boolean;
    grouping?: boolean;
    paging?: boolean;
    sorting?: boolean;
    resizing?: boolean;
    reordering?: boolean;
    columnMenu?: boolean;
    navigable?: boolean;
    pageSize?: number;
    height?: number;
    columns?: string;
    dataSource: DataSource;
    dataPath?: string;
    dataName: string;
}
