// @progress/kendo-angular-grid/esm2015/rendering/footer/footer.module.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { SharedModule } from "../../shared.module";
import { FooterComponent } from "./footer.component";
import * as i0 from "@angular/core";
const exportedModules = [
    FooterComponent
];
const importedModules = [
    CommonModule,
    SharedModule
];
/**
 * @hidden
 */
export class FooterModule {
    static exports() {
        return [];
    }
}
FooterModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FooterModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
FooterModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FooterModule, declarations: [FooterComponent], imports: [CommonModule,
        SharedModule], exports: [FooterComponent] });
FooterModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FooterModule, imports: [[...importedModules]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FooterModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [exportedModules],
                    exports: [exportedModules],
                    imports: [...importedModules]
                }]
        }] });
