// @progress/kendo-angular-inputs/esm2015/inputs.module.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { SliderModule } from './slider.module';
import { RangeSliderModule } from './rangeslider.module';
import { SwitchModule } from './switch.module';
import { NumericTextBoxModule } from './numerictextbox.module';
import { MaskedTextBoxModule } from './maskedtextbox.module';
import { TextBoxModule } from './textbox.module';
import { TextAreaModule } from './textarea.module';
import { CommonModule } from '@angular/common';
import { ColorPickerModule } from './colorpicker.module';
import { CheckBoxModule } from './checkbox.module';
import { RadioButtonModule } from './radiobutton.module';
import { FormFieldModule } from './formfield.module';
import { SignatureModule } from './signature.module';
import * as i0 from "@angular/core";
/**
 * Represents the [NgModule]({{ site.data.urls.angular['ngmoduleapi'] }})
 * definition for the Inputs components.
 *
 * @example
 *
 * ```ts-no-run
 * // Import the Inputs module
 * import { InputsModule } from '@progress/kendo-angular-inputs';
 *
 * // The browser platform with a compiler
 * import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
 * import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
 *
 * import { NgModule } from '@angular/core';
 *
 * // Import the app component
 * import { AppComponent } from './app.component';
 *
 * // Define the app module
 * _@NgModule({
 *     declarations: [AppComponent], // declare app component
 *     imports:      [BrowserModule, BrowserAnimationsModule, InputsModule], // import Inputs module
 *     bootstrap:    [AppComponent]
 * })
 * export class AppModule {}
 *
 * // Compile and launch the module
 * platformBrowserDynamic().bootstrapModule(AppModule);
 *
 * ```
 */
export class InputsModule {
}
InputsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: InputsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
InputsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: InputsModule, imports: [CommonModule], exports: [TextAreaModule,
        TextBoxModule,
        SliderModule,
        RangeSliderModule,
        SwitchModule,
        NumericTextBoxModule,
        MaskedTextBoxModule,
        ColorPickerModule,
        CheckBoxModule,
        RadioButtonModule,
        FormFieldModule,
        SignatureModule] });
InputsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: InputsModule, imports: [[CommonModule], TextAreaModule,
        TextBoxModule,
        SliderModule,
        RangeSliderModule,
        SwitchModule,
        NumericTextBoxModule,
        MaskedTextBoxModule,
        ColorPickerModule,
        CheckBoxModule,
        RadioButtonModule,
        FormFieldModule,
        SignatureModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: InputsModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [
                        TextAreaModule,
                        TextBoxModule,
                        SliderModule,
                        RangeSliderModule,
                        SwitchModule,
                        NumericTextBoxModule,
                        MaskedTextBoxModule,
                        ColorPickerModule,
                        CheckBoxModule,
                        RadioButtonModule,
                        FormFieldModule,
                        SignatureModule
                    ],
                    imports: [CommonModule]
                }]
        }] });
