// @progress/kendo-angular-inputs/esm2015/signature.module.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ButtonModule } from '@progress/kendo-angular-buttons';
import { DialogsModule } from '@progress/kendo-angular-dialog';
import { LocalizedSignatureMessagesDirective, SignatureCustomMessagesComponent } from './signature/localization';
import { SignatureComponent } from './signature/signature.component';
import * as i0 from "@angular/core";
/**
 * Represents the [NgModule]({{ site.data.urls.angular['ngmoduleapi'] }})
 * definition for the Signature component.
 *
 * @example
 *
 * ```ts-no-run
 * // Import the Signature module
 * import { SignatureModule } from '@progress/kendo-angular-inputs';
 *
 * // The browser platform with a compiler
 * import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
 *
 * import { NgModule } from '@angular/core';
 *
 * // Import the app component
 * import { AppComponent } from './app.component';
 *
 * // Define the app module
 * _@NgModule({
 *     declarations: [AppComponent], // declare app component
 *     imports:      [BrowserModule, SignatureModule], // import Signature module
 *     bootstrap:    [AppComponent]
 * })
 * export class AppModule {}
 *
 * // Compile and launch the module
 * platformBrowserDynamic().bootstrapModule(AppModule);
 *
 * ```
 */
export class SignatureModule {
}
SignatureModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: SignatureModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
SignatureModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: SignatureModule, declarations: [SignatureComponent,
        SignatureCustomMessagesComponent,
        LocalizedSignatureMessagesDirective], imports: [ButtonModule,
        CommonModule,
        DialogsModule], exports: [SignatureComponent,
        SignatureCustomMessagesComponent] });
SignatureModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: SignatureModule, imports: [[
            ButtonModule,
            CommonModule,
            DialogsModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: SignatureModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        SignatureComponent,
                        SignatureCustomMessagesComponent,
                        LocalizedSignatureMessagesDirective
                    ],
                    exports: [
                        SignatureComponent,
                        SignatureCustomMessagesComponent
                    ],
                    imports: [
                        ButtonModule,
                        CommonModule,
                        DialogsModule
                    ]
                }]
        }] });
