// @progress/kendo-angular-inputs/esm2015/textbox.module.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from "@angular/core";
import { TextBoxDirective } from "./textbox/textbox.directive";
import { TextBoxComponent } from "./textbox/textbox.component";
import { CommonModule } from "@angular/common";
import { EventsModule } from "@progress/kendo-angular-common";
import { TextBoxSuffixTemplateDirective } from "./textbox/textbox-suffix.directive";
import { TextBoxPrefixTemplateDirective } from './textbox/textbox-prefix.directive';
import { TextBoxCustomMessagesComponent } from './textbox/localization/custom-messages.component';
import { LocalizedTextBoxMessagesDirective } from './textbox/localization/localized-textbox-messages.directive';
import { SharedModule, SHARED_DIRECTIVES } from './shared.module';
import * as i0 from "@angular/core";
import * as i1 from "./shared/input-separator.component";
import * as i2 from "./shared/textarea.directive";
/**
 * Represents the [NgModule]({{ site.data.urls.angular['ngmoduleapi'] }})
 * definition for the TextBox directive.
 *
 * @example
 *
 * ```ts-no-run
 * // Import the TextBox module
 * import { TextBoxModule } from '@progress/kendo-angular-inputs';
 *
 * // The browser platform with a compiler
 * import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
 *
 * import { NgModule } from '@angular/core';
 *
 * // Import the app component
 * import { AppComponent } from './app.component';
 *
 * // Define the app module
 * _@NgModule({
 *     declarations: [AppComponent], // declare app component
 *     imports:      [BrowserModule, TextBoxModule], // import TextBox module
 *     bootstrap:    [AppComponent]
 * })
 * export class AppModule {}
 *
 * // Compile and launch the module
 * platformBrowserDynamic().bootstrapModule(AppModule);
 *
 * ```
 */
export class TextBoxModule {
}
TextBoxModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: TextBoxModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TextBoxModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: TextBoxModule, declarations: [TextBoxDirective,
        TextBoxComponent,
        TextBoxSuffixTemplateDirective,
        TextBoxPrefixTemplateDirective,
        TextBoxCustomMessagesComponent,
        LocalizedTextBoxMessagesDirective], imports: [CommonModule, EventsModule, SharedModule], exports: [TextBoxDirective,
        TextBoxComponent,
        TextBoxSuffixTemplateDirective,
        TextBoxPrefixTemplateDirective,
        EventsModule,
        TextBoxCustomMessagesComponent,
        LocalizedTextBoxMessagesDirective, i1.InputSeparatorComponent, i2.TextAreaDirective] });
TextBoxModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: TextBoxModule, imports: [[CommonModule, EventsModule, SharedModule], EventsModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: TextBoxModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        TextBoxDirective,
                        TextBoxComponent,
                        TextBoxSuffixTemplateDirective,
                        TextBoxPrefixTemplateDirective,
                        TextBoxCustomMessagesComponent,
                        LocalizedTextBoxMessagesDirective
                    ],
                    exports: [
                        TextBoxDirective,
                        TextBoxComponent,
                        TextBoxSuffixTemplateDirective,
                        TextBoxPrefixTemplateDirective,
                        EventsModule,
                        TextBoxCustomMessagesComponent,
                        LocalizedTextBoxMessagesDirective,
                        SHARED_DIRECTIVES
                    ],
                    imports: [CommonModule, EventsModule, SharedModule]
                }]
        }] });
