// @progress/kendo-angular-grid/esm2015/filtering/menu/filter-menu.module.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { FilterMenuComponent } from "./filter-menu.component";
import { FilterMenuContainerComponent } from "./filter-menu-container.component";
import { FilterMenuInputWrapperComponent } from "./filter-menu-input-wrapper.component";
import { StringFilterMenuInputComponent } from "./string-filter-menu-input.component";
import { StringFilterMenuComponent } from './string-filter-menu.component';
import { SharedFilterModule } from '../shared-filtering.module';
import { FilterMenuTemplateDirective } from './filter-menu-template.directive';
import { NumericFilterMenuComponent } from './numeric-filter-menu.component';
import { NumericFilterMenuInputComponent } from './numeric-filter-menu-input.component';
import { FilterMenuHostDirective } from './filter-menu-host.directive';
import { DateFilterMenuInputComponent } from './date-filter-menu-input.component';
import { DateFilterMenuComponent } from './date-filter-menu.component';
import { BooleanFilterMenuComponent } from './boolean-filter-menu.component';
import { FilterMenuDropDownListDirective } from './filter-menu-dropdownlist.directive';
import { BooleanFilterRadioButtonDirective } from './filter-radio-button.directive';
import * as i0 from "@angular/core";
const INTERNAL_COMPONENTS = [
    FilterMenuComponent,
    FilterMenuContainerComponent,
    FilterMenuInputWrapperComponent,
    StringFilterMenuInputComponent,
    StringFilterMenuComponent,
    FilterMenuTemplateDirective,
    NumericFilterMenuComponent,
    NumericFilterMenuInputComponent,
    DateFilterMenuInputComponent,
    DateFilterMenuComponent,
    FilterMenuHostDirective,
    BooleanFilterMenuComponent,
    FilterMenuDropDownListDirective,
    BooleanFilterRadioButtonDirective
];
const ENTRY_COMPONENTS = [
    StringFilterMenuComponent,
    NumericFilterMenuComponent,
    DateFilterMenuComponent,
    BooleanFilterMenuComponent
];
/**
 * @hidden
 */
export class FilterMenuModule {
    static exports() {
        return [
            StringFilterMenuComponent,
            FilterMenuTemplateDirective,
            NumericFilterMenuComponent,
            DateFilterMenuComponent,
            BooleanFilterMenuComponent,
            FilterMenuDropDownListDirective,
            BooleanFilterRadioButtonDirective,
            SharedFilterModule.exports()
        ];
    }
}
FilterMenuModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FilterMenuModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
FilterMenuModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FilterMenuModule, declarations: [FilterMenuComponent,
        FilterMenuContainerComponent,
        FilterMenuInputWrapperComponent,
        StringFilterMenuInputComponent,
        StringFilterMenuComponent,
        FilterMenuTemplateDirective,
        NumericFilterMenuComponent,
        NumericFilterMenuInputComponent,
        DateFilterMenuInputComponent,
        DateFilterMenuComponent,
        FilterMenuHostDirective,
        BooleanFilterMenuComponent,
        FilterMenuDropDownListDirective,
        BooleanFilterRadioButtonDirective], imports: [SharedFilterModule], exports: [FilterMenuComponent,
        FilterMenuContainerComponent,
        FilterMenuInputWrapperComponent,
        StringFilterMenuInputComponent,
        StringFilterMenuComponent,
        FilterMenuTemplateDirective,
        NumericFilterMenuComponent,
        NumericFilterMenuInputComponent,
        DateFilterMenuInputComponent,
        DateFilterMenuComponent,
        FilterMenuHostDirective,
        BooleanFilterMenuComponent,
        FilterMenuDropDownListDirective,
        BooleanFilterRadioButtonDirective, SharedFilterModule] });
FilterMenuModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FilterMenuModule, imports: [[SharedFilterModule], SharedFilterModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: FilterMenuModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [INTERNAL_COMPONENTS],
                    entryComponents: ENTRY_COMPONENTS,
                    exports: [INTERNAL_COMPONENTS, SharedFilterModule],
                    imports: [SharedFilterModule]
                }]
        }] });
