// @progress/kendo-schematics/src/collection/charts/circular/files/inline/__name@dasherize__/__name@dasherize__.component.ts
import { Component } from '@angular/core';<% if (mockedData) {%>
import { seriesData } from './data'; <% } %><% if (showLabels) {%>
import { SeriesLabelsContentArgs } from '@progress/kendo-angular-charts';<% } %><% if (legendTitleText || legendTitleAlign || legendTitlePosition) {%>
import { LegendTitle } from '@progress/kendo-angular-charts';<% } %>

@Component({
	selector: '<%= dasherize(name) %>-component',
	template: `
        <kendo-chart<% if (renderAs) {%>renderAs="<%=renderAs%>"<% } %>><% if (titleText) {%>
            <kendo-chart-title
                text="<%=titleText %>" <% if (titlePosition) {%>
                position="<%=titlePosition%>" <% } %><% if (titleAlign) {%>
                align="<%=titleAlign%>"<% } %>></kendo-chart-title><% }%><% if (legend) {%>
                <kendo-chart-legend<% if (legendAlign) {%>
                    align="<%=legendAlign%>"<% } %><% if (legendPosition) {%>
                    position="<%=legendPosition%>"<% } %><% if (legendOrientation) {%>
                    orientation="<%=legendOrientation%>"<% } %><% if (legendTitleText || legendTitleAlign || legendTitlePosition) {%>
                    [title]="legendTitle"<% } %>></kendo-chart-legend><% }%>
            <kendo-chart-series><% if (series.length) {%><% for (let item of series) {%>
                <kendo-chart-series-item<% if (item.categoryField) {%>
                    categoryField="<%=item.categoryField%>"<% } %><% if (item.currentField) {%>
                    currentField="<%=item.currentField%>"<% } %>
                    [data]="seriesData"<% if (item.field) {%>
                    field="<%=item.field%>"<% } %><% if (item.name) {%>
                    name="<%=item.name%>"<% } %><% if (item.type) {%>
                    type="<%=item.type%>"<% } %><% if (item.holeSize || item.holeSize === 0) {%>
                    [holeSize]="<%=item.holeSize%>"<% } %><% if (item.size || item.size === 0) {%>
                    [size]="<%=item.size%>"<% } %><% if (item.startAngle || item.startAngle === 0) {%>
                    [startAngle]="<%=item.startAngle%>"<% } %>><% if (showLabels) {%>
                    <kendo-chart-series-item-labels
                        [visible]="true"
                        [content]="labelContent"></kendo-chart-series-item-labels><% } %>
                </kendo-chart-series-item><% } %><% } %>
            </kendo-chart-series>
        </kendo-chart>
	`,<% if (inlineStyle) {%>
	styles: [``]<% } else {%>
	styleUrls: ['./<%= dasherize(name) %>.component.css']<% } %>
})
export class <%= classify(name) %>Component {
	<% if (mockedData) { %>public seriesData: any[] = seriesData;<%}%><% if (showLabels) {%>
    public labelContent: (args: SeriesLabelsContentArgs) => string = (args: SeriesLabelsContentArgs) => args.category;<% } %><% if (legendTitleText || legendTitleAlign || legendTitlePosition) {%>
    public legendTitle: LegendTitle = {<% if (legendTitleText) {%>
        text: '<%=legendTitleText%>',<% } %><% if (legendTitleAlign) {%>
        align: '<%=legendTitleAlign%>',<% } %><% if (legendTitlePosition) {%>
        position: '<%=legendTitlePosition%>'<% } %>
    };<% } %>
}
