// @progress/kendo-angular-dropdowns/esm2015/dropdowntrees/expanded-state/expand.directive.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Directive, Input } from '@angular/core';
import { ExpandDirective } from '@progress/kendo-angular-treeview';
import * as i0 from "@angular/core";
import * as i1 from "@progress/kendo-angular-treeview";
/**
 * A directive which manages the expanded state of the popup TreeView.
 */
export class DropDownTreesExpandDirective extends ExpandDirective {
    constructor(dropDownTree) {
        super(dropDownTree);
        this.dropDownTree = dropDownTree;
    }
    /**
     * @hidden
     *
     * Ensures a user-defined `isNodeExpanded` callback will not be overriden by the default directive setup.
     * Implemented as a value setter in the base directive, this just overrides the input name.
     */
    set isExpanded(value) {
        this.dropDownTree.isExpanded = value;
    }
    ;
}
DropDownTreesExpandDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: DropDownTreesExpandDirective, deps: [{ token: i1.ExpandableComponent }], target: i0.ɵɵFactoryTarget.Directive });
DropDownTreesExpandDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.16", type: DropDownTreesExpandDirective, selector: "[kendoDropDownTreeExpandable], [kendoMultiSelectTreeExpandable]", inputs: { isExpanded: ["isNodeExpanded", "isExpanded"] }, usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: DropDownTreesExpandDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[kendoDropDownTreeExpandable], [kendoMultiSelectTreeExpandable]'
                }]
        }], ctorParameters: function () { return [{ type: i1.ExpandableComponent }]; }, propDecorators: { isExpanded: [{
                type: Input,
                args: ['isNodeExpanded']
            }] } });
