// @progress/kendo-schematics/src/collection/charts/categorical/schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValueAxisType = exports.CategoryAxisType = exports.SeriesType = exports.DataSource = void 0;
var DataSource;
(function (DataSource) {
    /**
     * An example data source will be added to the project with predefined column definitions.
     * The data and column definitions can not be modified.
     *
     * This is the default.
     */
    DataSource["Example"] = "Example";
    /**
     * The user is expected to provide column definitions.
     * A mock data will be generated out of the column definitions.
     */
    DataSource["Mock Data"] = "Mock Data";
    /**
     * The user can specify a basic module to bind to and provide column definitions,
     * but in general the user will have to author some code to connect the data.
     */
    DataSource["Existing App Data"] = "Existing App Data";
})(DataSource = exports.DataSource || (exports.DataSource = {}));
var SeriesType;
(function (SeriesType) {
    SeriesType["area"] = "area";
    SeriesType["bar"] = "bar";
    SeriesType["bullet"] = "bullet";
    SeriesType["column"] = "column";
    SeriesType["horizontalWaterfall"] = "horizontalWaterfall";
    SeriesType["line"] = "line";
    SeriesType["radarArea"] = "radarArea";
    SeriesType["radarColumn"] = "radarColumn";
    SeriesType["radarLine"] = "radarLine";
    SeriesType["rangeArea"] = "rangeArea";
    SeriesType["rangeBar"] = "rangeBar";
    SeriesType["rangeColumn"] = "rangeColumn";
    SeriesType["waterfall"] = "waterfall";
})(SeriesType = exports.SeriesType || (exports.SeriesType = {}));
var CategoryAxisType;
(function (CategoryAxisType) {
    CategoryAxisType["category"] = "category";
    CategoryAxisType["date"] = "date";
})(CategoryAxisType = exports.CategoryAxisType || (exports.CategoryAxisType = {}));
var ValueAxisType;
(function (ValueAxisType) {
    ValueAxisType["numeric"] = "numeric";
    ValueAxisType["log"] = "log";
})(ValueAxisType = exports.ValueAxisType || (exports.ValueAxisType = {}));
