// @progress/kendo-schematics/src/collection/charts/categorical/files/inline/__name@dasherize__/__name@dasherize__.component.ts
import { Component } from '@angular/core';<% if (mockedData) {%>
import { seriesData } from './data'; <% } %><% if (legendTitleText || legendTitleAlign || legendTitlePosition) {%>
import { LegendTitle } from '@progress/kendo-angular-charts';<% } %>

@Component({
	selector: '<%= dasherize(name) %>-component',
	template: `
        <kendo-chart <% if (renderAs) {%>renderAs="<%=renderAs%>"<% } %>>
<% if (titleText) { %>
            <kendo-chart-title
                text="<%=titleText %>" <% if (titlePosition) {%>
                position="<%=titlePosition%>" <% } %><% if (titleAlign) {%>
                align="<%=titleAlign%>" <% } %>
                ></kendo-chart-title>
<% } %><% if (legend) { %>
            <kendo-chart-legend<% if (legendAlign) {%>
                align="<%=legendAlign%>"<% } %><% if (legendPosition) {%>
                position="<%=legendPosition%>"<% } %><% if (legendOrientation) {%>
                orientation="<%=legendOrientation%>"<% } %><% if (legendTitleText) {%>
                [title]="legendTitle"<% } %>></kendo-chart-legend>
<% } %>
            <kendo-chart-series>
<% for (let item of series) { %>
        <kendo-chart-series-item<% if (item.currentField) {%>
            currentField="<%=item.currentField%>"<% } %><% if (item.categoryField) {%>
            categoryField="<%=item.categoryField%>"<% } %>
            [data]="seriesData"<% if (item.field) {%>
            field="<%=item.field%>"<% } %><% if (item.name) {%>
            name="<%=item.name%>"<% } %><% if (item.type) {%>
            type="<%=item.type%>"<% } %><% if (item.width) {%>
            [width]="<%=item.width%>"<% } %>>
        </kendo-chart-series-item>
<% } %>
            </kendo-chart-series>
            <kendo-chart-category-axis>
<% for (let item of categoryAxis) { %>
                <kendo-chart-category-axis-item<% if (item.name) {%>
                    name="<%=item.name%>"<% } %><% if (item.type) {%>
                    type="<%=item.type%>"<% } %>><% if (item.title) {%>
                    <kendo-chart-category-axis-item-title text="<%=item.title%>"></kendo-chart-category-axis-item-title><% } %>
                </kendo-chart-category-axis-item>
<% } %>
            </kendo-chart-category-axis>
            <kendo-chart-value-axis>
<% for (let item of valueAxis) { %>
                <kendo-chart-value-axis-item<% if (item.name) {%>
                    name="<%=item.name%>"<% } %><% if (item.type) {%>
                    type="<%=item.type%>"<% } %>><% if (item.title) {%>
                    <kendo-chart-value-axis-item-title text="<%=item.title%>"></kendo-chart-value-axis-item-title><% } %>
                </kendo-chart-value-axis-item>
<% } %>
            </kendo-chart-value-axis>
        </kendo-chart>
	`,<% if (inlineStyle) {%>
	styles: [``]<% } else {%>
	styleUrls: ['./<%= dasherize(name) %>.component.css']<% } %>
})
export class <%= classify(name) %>Component {
	public seriesData: any[] = seriesData;<% if (legendTitleText || legendTitleAlign || legendTitlePosition) {%>
    public legendTitle: LegendTitle = {<% if (legendTitleText) {%>
        text: '<%=legendTitleText%>',<% } %><% if (legendTitleAlign) {%>
        align: '<%=legendTitleAlign%>',<% } %><% if (legendTitlePosition) {%>
        position: '<%=legendTitlePosition%>'<% } %>
    };<% } %>
}
