// @progress/kendo-angular-dateinputs/esm2015/dateinput/dateinput.module.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { DateInputComponent } from './dateinput.component';
import { CommonModule } from '@angular/common';
import { IntlModule } from '@progress/kendo-angular-intl';
import { DateInputLocalizedMessagesDirective } from './localization/dateinput-localized-messages.directive';
import { DateInputCustomMessagesComponent } from './localization/dateinput-custom-messages.component';
import { EventsModule } from '@progress/kendo-angular-common';
import * as i0 from "@angular/core";
/**
 * Represents the [NgModule](link:site.data.urls.angular['ngmoduleapi'])
 * definition for the DateInput component.
 */
export class DateInputModule {
}
DateInputModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: DateInputModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DateInputModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: DateInputModule, declarations: [DateInputComponent,
        DateInputCustomMessagesComponent,
        DateInputLocalizedMessagesDirective], imports: [CommonModule, IntlModule, EventsModule], exports: [DateInputComponent,
        DateInputCustomMessagesComponent,
        DateInputLocalizedMessagesDirective] });
DateInputModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: DateInputModule, imports: [[CommonModule, IntlModule, EventsModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: DateInputModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        DateInputComponent,
                        DateInputCustomMessagesComponent,
                        DateInputLocalizedMessagesDirective
                    ],
                    exports: [
                        DateInputComponent,
                        DateInputCustomMessagesComponent,
                        DateInputLocalizedMessagesDirective
                    ],
                    imports: [CommonModule, IntlModule, EventsModule]
                }]
        }] });
