// @progress/kendo-schematics/src/collection/scheduler/files/inline/__name@dasherize__/__name@dasherize__.component.ts
import { Component } from '@angular/core';
import { sampleData, displayDate } from './events-utc';
import { SchedulerEvent } from '@progress/kendo-angular-scheduler';

@Component({
    selector: '<%= dasherize(name) %>-component',
    template: `
        <kendo-scheduler [kendoSchedulerBinding]="events" [selectedDate]="selectedDate" scrollTime="08:00" style="height: 600px;">
<%
if (dayView) {
%>
            <kendo-scheduler-day-view> </kendo-scheduler-day-view>
<%
}
if (weekView) {
%>
            <kendo-scheduler-week-view> </kendo-scheduler-week-view>
<%
}
if (monthView) {
%>
            <kendo-scheduler-month-view> </kendo-scheduler-month-view>
<%
}
if (timelineView) {
%>
            <kendo-scheduler-timeline-view> </kendo-scheduler-timeline-view>
<%
}
if (agendaView) {
%>
            <kendo-scheduler-agenda-view> </kendo-scheduler-agenda-view>
<%
}
%>
        </kendo-scheduler>
        `,<% if (inlineStyle) {%>
    styles: [``]<% } else {%>
    styleUrls: ['./<%= dasherize(name) %>.component.css']<% } %>
})
export class <%= classify(name) %>Component  {
    public selectedDate: Date = displayDate;
    public events: SchedulerEvent[] = sampleData;
}
