// @progress/kendo-angular-dropdowns/esm2015/dropdowntrees/dropdowntrees.module.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { SharedModule } from '../common/shared.module';
import { TreeViewModule } from '@progress/kendo-angular-treeview';
import { SharedDirectivesModule } from '../common/shared-directives.module';
import { DropDownTreeComponent } from './dropdowntree.component';
import { MultiSelectTreeComponent } from './multiselecttree.component';
import { DropDownTreeFlatBindingDirective } from './data-binding/dropdowntree/flat-binding.directive';
import { DropDownTreeHierarchyBindingDirective } from './data-binding/dropdowntree/hierarchy-binding.directive';
import { MultiSelectTreeFlatBindingDirective } from './data-binding/multiselecttree/flat-binding.directive';
import { MultiSelectTreeHierarchyBindingDirective } from './data-binding/multiselecttree/hierarchy-binding.directive';
import { DropDownTreesExpandDirective } from './expanded-state/expand.directive';
import { NodeTemplateDirective } from './templates/node-template.directive';
import { MultiSelectTreeSummaryTagDirective } from './summary-tag/summary-tag.directive';
import * as i0 from "@angular/core";
const DROPDOWNTREE_DIRECTIVES = [
    DropDownTreeComponent,
    MultiSelectTreeComponent,
    DropDownTreeFlatBindingDirective,
    DropDownTreeHierarchyBindingDirective,
    MultiSelectTreeFlatBindingDirective,
    MultiSelectTreeHierarchyBindingDirective,
    DropDownTreesExpandDirective,
    NodeTemplateDirective,
    MultiSelectTreeSummaryTagDirective
];
/**
 * @hidden
 */
export class DropDownTreesModule {
}
DropDownTreesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: DropDownTreesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DropDownTreesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: DropDownTreesModule, declarations: [DropDownTreeComponent,
        MultiSelectTreeComponent,
        DropDownTreeFlatBindingDirective,
        DropDownTreeHierarchyBindingDirective,
        MultiSelectTreeFlatBindingDirective,
        MultiSelectTreeHierarchyBindingDirective,
        DropDownTreesExpandDirective,
        NodeTemplateDirective,
        MultiSelectTreeSummaryTagDirective], imports: [SharedModule, TreeViewModule], exports: [DropDownTreeComponent,
        MultiSelectTreeComponent,
        DropDownTreeFlatBindingDirective,
        DropDownTreeHierarchyBindingDirective,
        MultiSelectTreeFlatBindingDirective,
        MultiSelectTreeHierarchyBindingDirective,
        DropDownTreesExpandDirective,
        NodeTemplateDirective,
        MultiSelectTreeSummaryTagDirective, SharedDirectivesModule] });
DropDownTreesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: DropDownTreesModule, imports: [[SharedModule, TreeViewModule], SharedDirectivesModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: DropDownTreesModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [DROPDOWNTREE_DIRECTIVES],
                    exports: [DROPDOWNTREE_DIRECTIVES, SharedDirectivesModule],
                    imports: [SharedModule, TreeViewModule]
                }]
        }] });
