// @progress/kendo-angular-grid/esm2015/pager/pager-dropdown.directive.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Directive } from '@angular/core';
import { Keys } from '@progress/kendo-angular-common';
import * as i0 from "@angular/core";
import * as i1 from "@progress/kendo-angular-dropdowns";
/**
 * @hidden
 */
export class PagerDropDownListDirective {
    constructor(host) {
        this.host = host;
        this.keydownHandler = (e) => {
            if (e.keyCode === Keys.Escape && this.host.isOpen) {
                e.stopPropagation();
                this.host.toggle(false);
            }
        };
    }
    ngAfterViewInit() {
        const wrapperElement = this.host.wrapper.nativeElement;
        wrapperElement.addEventListener('keydown', this.keydownHandler, true);
    }
    ngOnDestroy() {
        this.host.wrapper.nativeElement.removeEventListener('keydown', this.keydownHandler);
    }
}
PagerDropDownListDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: PagerDropDownListDirective, deps: [{ token: i1.DropDownListComponent }], target: i0.ɵɵFactoryTarget.Directive });
PagerDropDownListDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.16", type: PagerDropDownListDirective, selector: "[kendoGridPagerDropDown]", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: PagerDropDownListDirective, decorators: [{
            type: Directive,
            args: [{ selector: '[kendoGridPagerDropDown]' }]
        }], ctorParameters: function () { return [{ type: i1.DropDownListComponent }]; } });
