// @progress/kendo-angular-dialog/esm2015/main.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2021 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
export { DialogComponent } from './dialog/dialog.component';
export { DialogTitleBarComponent } from './dialog/dialog-titlebar.component';
export { DialogContentBase } from './dialog/dialog-content-base';
export { DialogActionsComponent } from './dialog/dialog-actions.component';
export { DialogService } from './dialog/dialog.service';
export { DialogCloseResult, DialogRef, DialogSettings, DialogAction } from './dialog/models/';
export { WindowComponent } from './window/window.component';
export { WindowTitleBarComponent } from './window/window-titlebar.component';
export { WindowMaximizeActionDirective } from './window/actions/window-maximize-action.directive';
export { WindowMinimizeActionDirective } from './window/actions/window-minimize-action.directive';
export { WindowCloseActionDirective } from './window/actions/window-close-action.directive';
export { WindowRestoreActionDirective } from './window/actions/window-restore-action.directive';
export { WindowSettings, WindowRef, WindowCloseResult } from './window/models';
export { WindowService } from './window/window.service';
export { DragResizeService } from './window/drag-resize.service';
export { DialogModule } from './dialog.module';
export { WindowModule } from './window.module';
export { DialogsModule } from './dialogs.module';
// error NG3001: Unsupported private class
export { DialogContainerDirective } from './dialog/dialog-container.directive';
export { Messages } from './localization/messages';
export { LocalizedMessagesDirective } from './localization/localized-messages.directive';
export { CustomMessagesComponent } from './localization/custom-messages.component';
export { WindowContainerDirective } from './window/window-container.directive';
export { PreventableEvent } from './common/preventable-event';
